web/src/main/java/org/wso2/testgrid/web/api/SSOService.java

```
@@ -0,0 +1,153 @@
+
+    @javax.ws.rs.core.Context
+    ServletContext context;
+
+    @Context
+    HttpServletRequest request;
+
+    /**
+      * This has the implementation of the REST API for creating session and send the redirection information .
+      * @return Redirection to dashboard home page.
+     */
+    @POST
+    public Response createSession(@FormParam("SAMLResponse") String responseMessage) throws TestGridException {
+        XMLObject response;
+        org.opensaml.saml2.core.Response saml2Response;
+        try {
+            response = SSOAgentUtils.unmarshall(new String(Base64.decode(responseMessage), Charset.forName("UTF-8")));
+            saml2Response = (org.opensaml.saml2.core.Response) response;
+        } catch (SSOAgentException e) {
+            String msg = "Error occurred while unmarshalling SAMLResponse.";
+            logger.error(msg, e);
+            return Response.serverError().entity(
+                    new ErrorResponse.ErrorResponseBuilder().setMessage(msg + e.getMessage()).build()).build();
+        }
+
+        try {
+            validateSignature(saml2Response.getSignature());
+            request.getSession();
+            return Response.status(Response.Status.FOUND).
+                    header(HttpHeaders.LOCATION, Constants.WEBAPP_CONTEXT).
+                    type(MediaType.TEXT_PLAIN).build();
+        } catch (SSOAgentException e) {
+            String msg = "Observed an attempt with invalid SAMLResponse.";
+            logger.error(msg, e);
+            return Response.status(Response.Status.UNAUTHORIZED).entity(msg + e.getMessage()).build();
+        }
+    }
+
+    /**
+     * Validates the XML Signature object
+     *
+     * @param signature XMLObject
+     * @throws TestGridException if either SignatureProfile or Signature validation is failed.
+     */
+    private void validateSignature(XMLObject signature) throws SSOAgentException, TestGridException {
```

Shouldn't rather throw a specific exception here? Something like SignatureValidationFailureException?

---

Validator throws a 'ValidationException' already. But we are wrapping it with TestGridException and add line saying 'Signature validation failed'. Wouldn't that sufficient?

---

I think it's better to throw the ValidationException itself.